src/dune_pkg/opam_repo.ml

```diff
@@ -213,96 +213,127 @@ let scan_files_entries path =
-        Code_error.raise
-          "Enumerated file in directory but file can't be retrieved"
-          [ "local_file", Rev_store.File.to_dyn remote_file ]
-      | Some content ->
-        let local_file =
-          Path.Local.descendant ~of_:files_root remote_file_path |> Option.value_exn
-        in
-        { File_entry.local_file; original = Content content })
-;;
-
 module With_file = struct
+  type extra_files =
+    | Inside_files_dir
+    | Git_files of Rev_store.File.t list
+
   type nonrec t =
     { opam_file : OpamFile.OPAM.t
-    ; file : Path.t
+    ; package : OpamPackage.t
+    ; opam_file_path : Path.Local.t
     ; repo : t
+    ; extra_files : extra_files
     }
 
-  let file t = t.file
+  let file t =
+    match t.repo.source with
+    | Directory d -> Path.append_local d t.opam_file_path
+    | Repo _ ->
+      (* XXX fake path *)
+      Path.source @@ Path.Source.of_local t.opam_file_path
+  ;;
+
+  let package t = t.package
   let opam_file t = t.opam_file
   let repo t = t.repo
 end
 
-let load_opam_package_from_dir ~(dir : Path.t) opam_package =
-  Paths.opam_file opam_package
-  |> Path.append_local dir
+let load_opam_package_from_dir ~(dir : Path.t) package =
+  let opam_file_path = Paths.opam_file package in
+  Path.append_local dir opam_file_path
   |> if_exists
-  |> Option.map ~f:(fun opam_file_path ->
+  |> Option.map ~f:(fun opam_file_path_in_repo ->
     let opam_file =
-      Path.to_string opam_file_path
+      Path.to_string opam_file_path_in_repo
       |> OpamFilename.raw
       |> OpamFile.make
       |> OpamFile.OPAM.read
     in
     { With_file.opam_file
-    ; file = opam_file_path
+    ; package
+    ; opam_file_path
     ; repo = { source = Backend.Directory dir; serializable = None }
+    ; extra_files = With_file.Inside_files_dir
     })
 ;;
 
-(* Reads an opam package definition from an "opam" file in this repository
-   corresponding to a package (name and version). *)
-let load_opam_package t opam_package =
-  match t.source with
-  | Directory dir -> load_opam_package_from_dir ~dir opam_package |> Fiber.return
-  | Repo at_rev ->
-    let file = Paths.opam_file opam_package in
-    Rev_store.At_rev.content at_rev file
-    >>| Option.map ~f:(fun content ->
-      let opam_file =
-        (* the filename is used to read the version number *)
-        let filename =
-          Path.Local.to_string file |> OpamFilename.of_string |> OpamFile.make
+let get_opam_package_files with_files =
+  let indexed = List.mapi with_files ~f:(fun i w -> i, w) |> Int.Map.of_list_exn in
+  let from_dirs, from_git =
+    Int.Map.partition_map indexed ~f:(fun (with_file : With_file.t) ->
+      match with_file.extra_files with
+      | Git_files files -> Right (with_file.package, files)
+      | Inside_files_dir ->
+        let dir =
+          match with_file.repo.source with
+          | Directory root -> Path.append_local root (Paths.files_dir with_file.package)
+          | Repo _ -> assert false
```

Comment:
This assert looks rather ugly as it represents a case that shouldn't happen yet with the types modelled as it, it does. I would instead suggest adding an argument to this constructor that has the `root`, which would avoid this case.

(Or convert this assert to a `Code_error.raise` since there is definitely some bug if this is happening).

Reply:
I've converted this to an assert false. I have an idea on how to get rid of the assert completely, but it would take quite a bit more refactoring.